### DISCUSSION TITLE
Unauthenticated error on private nuget repository only when paginating

### DISCUSSION THREAD

**Discussion by platformoperations** [renovatebot/renovate#24601] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.102.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

For context: We're using Renovate to manage dependencies from a NuGet repository in a privately owned Artifactory setup. This worked for a long time, until one of the dependencies was not updated anymore to its latest version. Other dependencies from the same NuGet repository were still updated as normal, only one dependency was not found and as such not included in the MR/PR.

After looking into it, this only started to happen once the number of packages of that particular dependency exceeded 100. In the DEBUG logs we saw a line (added below) that a 401 Unauthorized status code was returned when attempting to find the package. The URL contains "skip=100" which according to [this API documentation](https://joelverhagen.github.io/NuGetUndocs/#odata-parameters) is to mean that the request is paginated.
I was able to confirm this locally when sending requests to our Artifactory.

The issue appears to be that authentication is not included in the request only in the case where we have more than 100 results and thus require pagination. Could this be correct?

--

A minimal reproduction is impossible for me as I have no access to a public artifact store / setup where I can reproduce this.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://<hostname>/artifactory/api/nuget/nuget-local/FindPackagesById()?%24select=Version%2CIsLatestVersion%2CProjectUrl%2CPublished&id=%27<package_name>%27&%24skip=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=38) (repository=<repository_name>)
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
Can you do this?

Set up a new reproduction repo with two dependencies from this registry - one which works and the one which doesn't.

Run Renovate on it with LOG_LEVEL=debug

Paste as much of the log possible to a gist and link here

**Reply by platformoperations** [reply]
I've created a gist [here](https://gist.github.com/platformoperations/de4b8df3088c612db1e29eeb78d6023d#file-renovate-debug-logs-txt).

`Utilities.Monitoring` is the one that does not work
`Utilities.HealthCheck` does

Let me know if I can provide any additional information!

**Reply by rarkins** [reply]
These http logs are interesting:

```
         "https://artifactory.jfrog.io/artifactory/api/nuget/nuget-local/FindPackagesById() (GET,401)": 1,
         "https://artifactory.jfrog.io/base/api/nuget/v3/nuget-local/FindPackagesById() (GET,200)": 2
```

Note 2 requests to the second one and one to the first. The first is the one which failed earlier:

```
DEBUG: GET https://artifactory.jfrog.io/artifactory/api/nuget/nuget-local/FindPackagesById()?%24select=Version%2CIsLatestVersion%2CProjectUrl%2CPublished&id=%27Utilities.Monitoring%27&%24skip=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=64) (repository=renovate-problem)
```

My guess is that when Renovate calls `https://artifactory.jfrog.io/base/api/nuget/v3/nuget-local/FindPackagesById()` then gets the other URL from this logic:

```
    const nextPkgUrlListLink = pkgVersionsListDoc
      .childrenNamed('link')
      .find((node) => node.attr.rel === 'next');

    pkgUrlList = nextPkgUrlListLink ? nextPkgUrlListLink.attr.href : null;
```

You would need to `curl` the original URL in verbose mode to be able to confirm that.

**Reply by platformoperations** [reply]
That seems correct!

I can confirm that when I locally query 
```
https://artifactory.jfrog.io/base/api/nuget/v3/nuget-local/FindPackagesById()?$select=Version,IsLatestVersion,ProjectUrl,Published&id='Utilities.Monitoring'
```

I receive the following in next: 

```
<link rel="next" href="https://artifactory.jfrog.io/artifactory/api/nuget/nuget-local/FindPackagesById()?%24select=Version%2CIsLatestVersion%2CProjectUrl%2CPublished&amp;id=%Utilities.Monitoring%27&amp;%24skip=100"/>
```

which is quite unexpected to me! I wonder if this is a nuget quirk? I don't know where it would be looking to build up that URL.

The v2 URL is not included in the hostRules, which should then explain the 401 I guess?

**Reply by rarkins** [reply]
What happens if you try adding that to hostRules?
